File: test/language/templates/cpp/StructTemplatedTypeArgumentTest.cpp
#include "gtest/gtest.h"

#include "templates/struct_templated_type_argument/StructTemplatedTypeArgument.h"

#include "zserio/RebindAlloc.h"

namespace templates
{
namespace struct_templated_type_argument
{

using allocator_type = StructTemplatedTypeArgument::allocator_type;
using string_type = zserio::string<allocator_type>;

TEST(StructTemplatedTypeArgumentTest, readWrite)
{
    StructTemplatedTypeArgument structTemplatedTypeArgument;
    structTemplatedTypeArgument.setParamHolder(ParamHolder_uint32{42});
    structTemplatedTypeArgument.setParameterized(Parameterized_uint32{string_type{"description"}, 13});
    structTemplatedTypeArgument.initializeChildren();
    structTemplatedTypeArgument.initializeOffsets();

    zserio::BitBuffer bitBuffer = zserio::BitBuffer(1024 * 8);
    zserio::BitStreamWriter writer(bitBuffer);
    structTemplatedTypeArgument.write(writer);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    StructTemplatedTypeArgument readStructTemplatedTypeArgument(reader);

    ASSERT_TRUE(structTemplatedTypeArgument == readStructTemplatedTypeArgument);
}

} // namespace struct_templated_type_argument
} // namespace templates

File: test/language/templates/python/StructTemplatedTypeArgumentTest.py
import unittest
import zserio

from testutils import getZserioApi

class StructTemplatedTypeArgumentTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.api = getZserioApi(__file__, "templates.zs").struct_templated_type_argument

    def testReadWrite(self):
        paramHolder = self.api.ParamHolder_uint32(42)
        structTemplatedTypeArgument = self.api.StructTemplatedTypeArgument(
            paramHolder,
            self.api.Parameterized_uint32(paramHolder, "description", 13)
        )

        writer = zserio.BitStreamWriter()
        structTemplatedTypeArgument.initialize_offsets(writer.bitposition)
        structTemplatedTypeArgument.write(writer)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        readStructTemplatedTypeArgument = self.api.StructTemplatedTypeArgument()
        readStructTemplatedTypeArgument.read(reader)
        self.assertEqual(structTemplatedTypeArgument, readStructTemplatedTypeArgument)
